Implement policies based on chunk creation time

pdipesh02 opened the pull request:
- The idea with these changes is to support INTERVAL boundaries for
  INTEGER type columns in policy APIs. The boundary calculation will
  be done using chunk creation time metadata.
- Added a new index on hypertable_id and creation_time on timescaledb
  catalog table CHUNK.
- Added access methods to perform index scan on CHUNK table using the
  index created on hypertable_id and creation_time.
- Added ChunkVec to accumulated the chunks satisfying the range
  criteria during index scan.
- Updated show_chunks, drop_chunks and retention policy APIs to get the
  affected chunks using chunk creation time metadata based on the
  INTERVAL boundary specified for the INTEGER columns.

pdipesh02 committed:
Add metadata for chunk creation time

- Added creation_time attribute to _timescaledb_catalog's table "chunk".
  Also, updated corresponding view timescaledb_information.chunks to
  include chunk_creation_time attribute.
- Chunk creation time for new chunks is calculated using function
  "GetCurrentTimestamp()" during chunk creation to handle new
  partitioning range (time or serial/bigserial/int/bigint etc...).
- Added changes to handle upgrade scenarios where chunk creation time is
  assigned to old chunks using now() function during upgrade.
- All relevant queries/views/test output has been updated accordingly.

fabriziomello commented on src/chunk.c:
```diff
@@ -4871,3 +4875,9 @@ ts_chunk_get_osm_chunk_id(int hypertable_id)
 
 	return chunk_id;
 }
+
+TimestampTz
+ts_chunk_get_creation_time(Oid chunk_oid)
+{
+	return GetCurrentTimestamp();
+}
```
Leftover?

pdipesh02 replied:
Yes, this needs to integrated with follow up PR on policy implementation.

fabriziomello commented on sql/updates/latest-dev.sql:
```diff
@@ -0,0 +1,92 @@
+-- add a new column creation_time to chunk catalog table
+ALTER TABLE _timescaledb_catalog.chunk ADD COLUMN creation_time TIMESTAMPTZ;
+UPDATE _timescaledb_catalog.chunk SET creation_time = now();
```
I guess we can do a better assumption here using for example `pg_stat_file` function, something like:

```suggestion
UPDATE
    _timescaledb_catalog.chunk c
SET
    creation_time = (pg_catalog.pg_stat_file(pg_catalog.pg_relation_filepath(r.oid))).change
FROM
    pg_class r, pg_namespace n
WHERE
    r.relnamespace = n.oid
    AND r.relname = c.table_name
    AND n.nspname = c.schema_name
    AND r.relkind = 'r'
    AND c.dropped IS FALSE;
UPDATE _timescaledb_catalog.chunk SET creation_time = now() WHERE creation_time IS NULL;
```
The second UPDATE is just to make sure we'll don't left any record without creation time information. One improvement is try to use pg_stat_file.creation field when the update run on Windows platform. https://www.postgresql.org/docs/15/functions-admin.html

pdipesh02 committed:
Implement policies based on chunk creation time

- The idea with these changes is to support INTERVAL boundaries for
  INTEGER type columns in policy APIs. The boundary calculation will
  be done using chunk creation time metadata.
- Added a new index on hypertable_id and creation_time on timescaledb
  catalog table CHUNK.
- Added access methods to perform index scan on CHUNK table using the
  index created on hypertable_id and creation_time.
- Added ChunkVec to accumulated the chunks satisfying the range
  criteria during index scan.
- Updated show_chunks, drop_chunks and retention policy APIs to get the
  affected chunks using chunk creation time metadata based on the
  INTERVAL boundary specified for the INTEGER columns.

nikkhils commented:
This is being further tracked via https://github.com/timescale/timescaledb/pull/6177
